discovery/watcher.go

```diff
@@ -44,19 +44,22 @@ type InitState struct {
 type SubscribeState struct{}
 
 type Watcher struct {
+	ctx       context.Context
```

Comment:
You may find that linters complain about this because of the general [guidance](https://go.dev/blog/context-and-structs) around not making contexts struct fields.

I think this is bit of an exception to the rule, because it's a long-lived component whos lifetime is tied to the application rather than individual request/connection/whatever. Also, gRPC's `ClientConn` hangs on to the context in this way anyway 🤷🏻 

I think it's fine!